File: src/prod/src/managed/Api/src/System/Fabric/health/ApplicationHealthStateChunkList.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric.Health
{
    using Interop;
    using System.Collections.Generic;

    /// <summary>
    /// Represents a list that contains <see cref="System.Fabric.Health.ApplicationHealthStateChunk"/> items.
    /// </summary>
    /// <remarks>The list can be obtained through health state chunk queries. The queries may have as result more chunks that can fit a message.
    /// In this case, the list of items that fit the message is returned plus a count that shows how many total items are available.</remarks>
    public sealed class ApplicationHealthStateChunkList : HealthStateChunkList<ApplicationHealthStateChunk>
    {
        /// <summary>
        /// Instantiates an empty <see cref="System.Fabric.Health.ApplicationHealthStateChunkList"/>.
        /// </summary>
        public ApplicationHealthStateChunkList() : base()
        {
        }

        internal ApplicationHealthStateChunkList(IList<ApplicationHealthStateChunk> list) : base(list)
        {
        }

        internal static unsafe ApplicationHealthStateChunkList CreateFromNativeList(IntPtr nativeListPtr)
        {
            var retval = new ApplicationHealthStateChunkList();

            if (nativeListPtr != null)
            {
                var nativeList = (NativeTypes.FABRIC_APPLICATION_HEALTH_STATE_CHUNK_LIST*)nativeListPtr;

                var nativeItemArray = (NativeTypes.FABRIC_APPLICATION_HEALTH_STATE_CHUNK*)nativeList->Items;
                for (int i = 0; i < nativeList->Count; ++i)
                {
                    var nativeItem = *(nativeItemArray + i);
                    retval.Add(ApplicationHealthStateChunk.FromNative(nativeItem));
                }

                retval.TotalCount = (long)nativeList->TotalCount;
            }
            
            return retval;
        }
    }
}

File: src/prod/src/ServiceModel/ApplicationHealthStateChunkList.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace std;
using namespace Common;
using namespace ServiceModel;

INITIALIZE_SIZE_ESTIMATION(ApplicationHealthStateChunkList)

StringLiteral const TraceSource("ApplicationHealthStateChunkList");

ApplicationHealthStateChunkList::ApplicationHealthStateChunkList()
    : HealthStateChunkList()
    , items_()
{
}

ApplicationHealthStateChunkList::ApplicationHealthStateChunkList(
    ULONG totalCount,
    std::vector<ApplicationHealthStateChunk> && items)
    : HealthStateChunkList(totalCount)
    , items_(move(items))
{
}

ApplicationHealthStateChunkList::~ApplicationHealthStateChunkList()
{
}

ErrorCode ApplicationHealthStateChunkList::ToPublicApi(
    __in Common::ScopedHeap & heap,
    __out FABRIC_APPLICATION_HEALTH_STATE_CHUNK_LIST & publicApplicationHealthStateChunkList) const
{
    publicApplicationHealthStateChunkList.TotalCount = this->TotalCount;
    auto error = PublicApiHelper::ToPublicApiList<ApplicationHealthStateChunk, FABRIC_APPLICATION_HEALTH_STATE_CHUNK, FABRIC_APPLICATION_HEALTH_STATE_CHUNK_LIST>(
        heap,
        items_,
        publicApplicationHealthStateChunkList);
    if (!error.IsSuccess()) { return error; }

    return ErrorCode::Success();
}

Common::ErrorCode ApplicationHealthStateChunkList::FromPublicApi(FABRIC_APPLICATION_HEALTH_STATE_CHUNK_LIST const * publicApplicationHealthStateChunkList)
{
    if (publicApplicationHealthStateChunkList == NULL)
    {
        return ErrorCode::Success();
    }

    this->TotalCount = publicApplicationHealthStateChunkList->TotalCount;
    auto error = PublicApiHelper::FromPublicApiList<ApplicationHealthStateChunk, FABRIC_APPLICATION_HEALTH_STATE_CHUNK_LIST>(publicApplicationHealthStateChunkList, items_);
    if (!error.IsSuccess()) { return error; }

    return ErrorCode::Success();
}
